### ISSUE TITLE
.only not working in 3.0.2

### ISSUE THREAD

**Issue by richardlay** [mochajs/mocha#2430]
We run mocha tests using grunt with a simple config:
```
mochaTest: {
            functional: {
                options: {
                    reporter: 'dot',
                    timeout: '10000',
                    recursive: true
                },
                src: ['tests/functional/**/test*.js']
            },
            unit: {
                options: {
                    reporter: 'dot',
                    timeout: '10000',
                    recursive: true
                },
                src: ['tests/unit/**/test*.js']
            }
        },
```

However when we upgraded to mocha 3.0.2, the .only feature does not work anymore. It runs all tests all the time. It works fine with mocha 2.5.3.

I tested it using mocha directly using:
```
mocha tests/functional/**/test*.js
```
and .only still not working.

Any help appreciated.

**Comment by gurdiga** [comment]
Hey @richardlay! 🙂

[I have tried to reproduce this](https://asciinema.org/a/c78cub6orwypmklok9v4ggw5q) with a fresh install and failed. 😶

Could you please take a quick look and maybe spot what’s different in my scenario from yours, and maybe you can come up with a small subset of your scenario that proves the misbehavior. 🤓

**Comment by richardlay** [comment]
Hi @gurdiga 

After doing some simple tests, I could not reproduce it either using the mocha command line.
Actually I don't think my mocha command I posted before is valid... don't think you can specify a regex path like that.

However, grunt-mocha-test accepts a regex path, and it looks like it binds each file to mocha. 
grunt-mocha-test MochaWrapper.js snippet:
```
  params.files.forEach(function(file) {
    file.src.forEach(mocha.addFile.bind(mocha));
  });
```
Could that be the issue?

**Comment by boneskull** [comment]
Not sure what `grunt-mocha-test` is up to.  They may need to change something for 3.x; I'd suggest opening an issue there, if you haven't already.  The programmatic API isn't considered "public", so I'm not surprised things are breaking.

**Comment by gurdiga** [comment]
@richardlay Maybe you can prepare a repo with the smallest setup that reproduces the issue. I’m open to digging into it some more to find what’s wrong. 🤓

What do you think? 🤔

**Comment by Turbo87** [comment]
@boneskull I'm currently working on getting `ember-cli-mocha` updated to Mocha 3. Everything seems to be working fine except for this issue here. I'll try to extract a small reproduction case.

**Comment by Turbo87** [comment]
nevermind, it seems that our issue is most likely related to https://github.com/teddyzeenny/ember-mocha-adapter/blob/master/adapter.js#L169-L172 and not a bug in Mocha.

**Comment by richardlay** [comment]
@gurdiga Hi, sorry taken a while to get back to you. Here's a setup that reproduces the issue. Just run "npm test" (assuming you have grunt installed). Many thanks.

https://github.com/richardlay/grunt-mocha-302

**Comment by pghalliday** [comment]
I'm picking this up and know the cause in `grunt-mocha-test` - unfortunately in order to handle asynchronous errors correctly within grunt I have to replicate the procedure in `mocha.run`, split it in 2 and wrap the call to `runner.run` in a domain. This procedure has changed and I just need to update my implementation (copy).

As an aside, if this were split into 2 stages within the mocha API it would make my life easier :) (Or if  it were promise based, which makes catching asynchronous errors much simpler)

**Comment by pghalliday** [comment]
This is now fixed in version 0.13.0 of grunt-mocha-test. I'm also working on a pull request for mocha that will hopefully prevent breaking changes when mocha is updated

**Comment by pghalliday** [comment]
Didn't need a pull request - mocha seems to have stopped leaking unhandled exceptions asynchronously :) As such I have removed my workaround and now use the API as documented
